### PR TITLE
Sketch out Igalia's testing strategy

### PR DESCRIPTION
**Pull request by kbara**

### CODE REVIEW

File: src/program/lwaftr/doc/testing.md [hunk 1]
```diff
@@ -0,0 +1,46 @@
+# Testing
+
+The lwAFTR and associated software is tested in several ways. Igalia uses unit
+tests, integration tests ("selftest.sh"), end-to-end tests, and does an
+increasing amount of automated performance analysis with Hydra.
+
+An example of our testing is how ARP support is tested. ARP is implemented in a
+small app, independently of the lwAFTR, and included with the lwAFTR an app
```
Comment (teknico) [inline src/program/lwaftr/doc/testing.md]: This sentence doesn't flow that well, it's hard to understand.

File: src/program/lwaftr/doc/testing.md [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+whole app network, and make sure that the end result is byte for byte identical
+with what is expected. Both the implementation and the testing were done with
+careful attention paid to RFC 826 throughout, as well as to dumps of ARP packets
+from the live network the developer was on.
+
+ARP has two main parts: resolving a remote address, and providing the address of
+the lwAFTR on request. The first one can be tested in a network by specifying
+only the remote IP of the next IPv4 hop (not the ethernet address), then sending
+packets through the lwAFTR and confirming on the remote host that they are
+arriving. The latter can be tested by issuing an ARP request to the lwaftr from
```
Comment (teknico) [inline src/program/lwaftr/doc/testing.md]: `lwaftr` -> `lwAFTR`

File: src/program/lwaftr/doc/testing.md [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+new entry corresponding to the lwAFTR.  The end to end tests simulate both of
+these cases, but with captured packets rather than a live network.
+
+There are a hierarchy of tests. Unit tests are internally orientedf, and make
+sure that basic functionality and some edge cases are tested. By convention,
+they are found in functions called selftest().
+
+Integration tests tend to be called selftest.sh, or invoked by selftesh.sh.
+These test larger components of the system. The end to end tests have IPv4 and
+IPv6 packet captures, run them through the lwaftr app network, and compare the
```
Comment (teknico) [inline src/program/lwaftr/doc/testing.md]: `lwaftr` -> `lwAFTR`

File: src/program/lwaftr/doc/testing.md [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+ARP has two main parts: resolving a remote address, and providing the address of
+the lwAFTR on request. The first one can be tested in a network by specifying
+only the remote IP of the next IPv4 hop (not the ethernet address), then sending
+packets through the lwAFTR and confirming on the remote host that they are
+arriving. The latter can be tested by issuing an ARP request to the lwaftr from
+another machine; if the other machine runs Linux, `arp -n` should then show a
+new entry corresponding to the lwAFTR.  The end to end tests simulate both of
+these cases, but with captured packets rather than a live network.
+
+There are a hierarchy of tests. Unit tests are internally orientedf, and make
```
Comment (teknico) [inline src/program/lwaftr/doc/testing.md]: `orientedf` -> `oriented`

**Review by teknico — APPROVED** [review]
LGTM with a few minor changes.
